Deploy more smart pointers in SelectorQuery.cpp

AlexsanderDamaceno opened the pull request:
#### 09c341383ede98c7e50b0eefffb9081e9db45986
<pre>
Deploy more smart pointers in SelectorQuery.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=278848">https://bugs.webkit.org/show_bug.cgi?id=278848</a>

Reviewed by Chris Dumez.

Deploy more smart pointers in SelectorQuery.cpp

* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::SelectorDataList::closest const):
(WebCore::SelectorDataList::executeFastPathForIdSelector const):
(WebCore::elementsForLocalName):
(WebCore::anyElement):
(WebCore::SelectorDataList::executeSingleTagNameSelectorData const):
(WebCore::SelectorDataList::executeSingleClassNameSelectorData const):
(WebCore::SelectorDataList::executeSingleAttributeExactSelectorData const):
(WebCore::SelectorDataList::executeSingleSelectorData const):
(WebCore::SelectorDataList::executeSingleMultiSelectorData const):
(WebCore::SelectorDataList::executeCompiledSimpleSelectorChecker const):
(WebCore::SelectorDataList::executeCompiledSelectorCheckerWithCheckingContext const):
(WebCore::SelectorDataList::executeCompiledSingleMultiSelectorData const):

Canonical link: <a href="https://commits.webkit.org/283123@main">https://commits.webkit.org/283123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b72977b689de717b341ab8c9dc531e0691dd0735

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64670 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44035 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17266 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68693 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15276 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66787 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51809 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15555 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52005 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10538 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67736 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55952 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32627 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13328 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14149 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13658 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70400 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8615 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13159 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59331 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8649 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56038 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59511 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7119 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/784 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9901 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39847 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40924 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42107 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40668 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/SelectorQuery.cpp:
```diff
@@ -199,10 +199,10 @@ bool SelectorDataList::matches(Element& targetElement) const
         for (auto& selector : m_selectors) {
-            if (auto* candidateElement = selectorClosest(selector, currentElement, targetElement))
-                return candidateElement;
+            if (RefPtr candidateElement = selectorClosest(selector, currentElement.get(), targetElement))
```
The RefPtr doesn't seem helpful here?

cdumez commented on Source/WebCore/dom/SelectorQuery.cpp:
```diff
@@ -251,8 +251,8 @@ ALWAYS_INLINE void SelectorDataList::executeFastPathForIdSelector(const Containe
         for (auto& element : *elements) {
-            if ((rootNodeIsTreeScopeRoot || element->isDescendantOf(rootNode)) && selectorMatches(selectorData, element, rootNode)) {
-                appendOutputForElement(output, element);
+            if ((rootNodeIsTreeScopeRoot || element->isDescendantOf(rootNode)) && selectorMatches(selectorData, element.get(), rootNode)) {
```
Why do we need the extra `.get()` here?

cdumez commented on Source/WebCore/dom/SelectorQuery.cpp:
```diff
@@ -251,8 +251,8 @@ ALWAYS_INLINE void SelectorDataList::executeFastPathForIdSelector(const Containe
-            if ((rootNodeIsTreeScopeRoot || element->isDescendantOf(rootNode)) && selectorMatches(selectorData, element, rootNode)) {
-                appendOutputForElement(output, element);
+            if ((rootNodeIsTreeScopeRoot || element->isDescendantOf(rootNode)) && selectorMatches(selectorData, element.get(), rootNode)) {
+                appendOutputForElement(output, element.get());
```
ditto.

cdumez commented on Source/WebCore/dom/SelectorQuery.cpp:
```diff
@@ -326,17 +326,17 @@ static inline void elementsForLocalName(const ContainerNode& rootNode, const Ato
-                appendOutputForElement(output, element);
+        for (Ref element : descendantsOfType<Element>(const_cast<ContainerNode&>(rootNode))) {
+            if (element->tagQName().localName() == localName) {
+                appendOutputForElement(output, element.get());
```
do we really need `.get()` here? Same comment in many other places in this PR.

AlexsanderDamaceno replied:
If not put get, it pass the smart pointer object for the function and do not compile.

cdumez replied:
`Ref<T>` can get implicitly converted into a `T&`, so I don't get why it is needed for compilation.

Also, in this patch, you added `.get()` in one place where the variable type didn't seem to change.

AlexsanderDamaceno replied:
you're right, I was doing wrong assumption here, thank you.

cdumez commented on Source/WebCore/dom/SelectorQuery.cpp:
```diff
@@ -479,11 +479,11 @@ ALWAYS_INLINE void SelectorDataList::executeSingleMultiSelectorData(const Contai
 
-        if (selectorChecker(&element)) {
-            appendOutputForElement(output, element);
+        if (selectorChecker(&element.get())) {
```
`element.ptr()`

cdumez commented on Source/WebCore/dom/SelectorQuery.cpp:
```diff
@@ -496,11 +496,11 @@ ALWAYS_INLINE void SelectorDataList::executeCompiledSelectorCheckerWithCheckingC
 
-        if (selectorChecker(&element, &checkingContext)) {
-            appendOutputForElement(output, element);
+        if (selectorChecker(&element.get(), &checkingContext)) {
```
`element.ptr()`

cdumez commented on Source/WebCore/dom/SelectorQuery.cpp:
```diff
@@ -512,19 +512,19 @@ ALWAYS_INLINE void SelectorDataList::executeCompiledSingleMultiSelectorData(cons
             bool matched = false;
             if (selector.compiledSelector.status == SelectorCompilationStatus::SimpleSelectorChecker)
-                matched = SelectorCompiler::querySelectorSimpleSelectorChecker(selector.compiledSelector, &element);
+                matched = SelectorCompiler::querySelectorSimpleSelectorChecker(selector.compiledSelector, &element.get());
```
`element.ptr()`

cdumez commented on Source/WebCore/dom/SelectorQuery.cpp:
```diff
@@ -512,19 +512,19 @@ ALWAYS_INLINE void SelectorDataList::executeCompiledSingleMultiSelectorData(cons
             else {
                 ASSERT(selector.compiledSelector.status == SelectorCompilationStatus::SelectorCheckerWithCheckingContext);
-                matched = SelectorCompiler::querySelectorSelectorCheckerWithCheckingContext(selector.compiledSelector, &element, &checkingContext);
+                matched = SelectorCompiler::querySelectorSelectorCheckerWithCheckingContext(selector.compiledSelector, &element.get(), &checkingContext);
```
`element.ptr()`

AlexsanderDamaceno committed:
Deploy more smart pointers in SelectorQuery.cpp
https://bugs.webkit.org/show_bug.cgi?id=278848

Reviewed by Chris Dumez.

Deploy more smart pointers in SelectorQuery.cpp

* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::SelectorDataList::closest const):
(WebCore::SelectorDataList::executeFastPathForIdSelector const):
(WebCore::elementsForLocalName):
(WebCore::anyElement):
(WebCore::SelectorDataList::executeSingleTagNameSelectorData const):
(WebCore::SelectorDataList::executeSingleClassNameSelectorData const):
(WebCore::SelectorDataList::executeSingleAttributeExactSelectorData const):
(WebCore::SelectorDataList::executeSingleSelectorData const):
(WebCore::SelectorDataList::executeSingleMultiSelectorData const):
(WebCore::SelectorDataList::executeCompiledSimpleSelectorChecker const):
(WebCore::SelectorDataList::executeCompiledSelectorCheckerWithCheckingContext const):
(WebCore::SelectorDataList::executeCompiledSingleMultiSelectorData const):

Canonical link: https://commits.webkit.org/283123@main
